File: stripe/src/test/java/com/stripe/android/model/wallets/WalletFactoryTest.java
package com.stripe.android.model.wallets;

import android.os.Parcel;

import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

@RunWith(RobolectricTestRunner.class)
public class WalletFactoryTest {
    private static final String VISA_WALLET_JSON = "{\n" +
            "\t\"type\": \"visa_checkout\",\n" +
            "\t\"dynamic_last4\": \"1234\",\n" +
            "\t\"visa_checkout\": {\n" +
            "\t\t\"billing_address\": {\n" +
            "\t\t\t\"city\": \"San Francisco\",\n" +
            "\t\t\t\"country\": \"US\",\n" +
            "\t\t\t\"line1\": \"510 Townsend St\",\n" +
            "\t\t\t\"postal_code\": \"94103\",\n" +
            "\t\t\t\"state\": \"CA\"\n" +
            "\t\t},\n" +
            "\t\t\"email\": \"me@example.com\",\n" +
            "\t\t\"name\": \"John Doe\",\n" +
            "\t\t\"shipping_address\": {\n" +
            "\t\t\t\"city\": \"San Francisco\",\n" +
            "\t\t\t\"country\": \"US\",\n" +
            "\t\t\t\"line1\": \"1355 Market St\",\n" +
            "\t\t\t\"postal_code\": \"94103\",\n" +
            "\t\t\t\"state\": \"CA\"\n" +
            "\t\t}\n" +
            "\t}\n" +
            "}";

    private static final String MASTERPASS_WALLET_JSON = "{\n" +
            "\t\"type\": \"master_pass\",\n" +
            "\t\"dynamic_last4\": \"1234\",\n" +
            "\t\"master_pass\": {\n" +
            "\t\t\"billing_address\": {\n" +
            "\t\t\t\"city\": \"San Francisco\",\n" +
            "\t\t\t\"country\": \"US\",\n" +
            "\t\t\t\"line1\": \"510 Townsend St\",\n" +
            "\t\t\t\"postal_code\": \"94103\",\n" +
            "\t\t\t\"state\": \"CA\"\n" +
            "\t\t},\n" +
            "\t\t\"email\": \"me@example.com\",\n" +
            "\t\t\"name\": \"John Doe\",\n" +
            "\t\t\"shipping_address\": {\n" +
            "\t\t\t\"city\": \"San Francisco\",\n" +
            "\t\t\t\"country\": \"US\",\n" +
            "\t\t\t\"line1\": \"1355 Market St\",\n" +
            "\t\t\t\"postal_code\": \"94103\",\n" +
            "\t\t\t\"state\": \"CA\"\n" +
            "\t\t}\n" +
            "\t}\n" +
            "}";

    private static final String AMEX_EXPRESS_CHECKOUT_WALLET_JSON = "{\n" +
            "\t\"type\": \"amex_express_checkout\",\n" +
            "\t\"dynamic_last4\": \"1234\",\n" +
            "\t\"amex_express_checkout\": {}\n" +
            "}";

    private static final String APPLE_PAY_WALLET_JSON = "{\n" +
            "\t\"type\": \"apple_pay\",\n" +
            "\t\"dynamic_last4\": \"1234\",\n" +
            "\t\"apple_pay\": {}\n" +
            "}";

    private static final String GOOGLE_PAY_WALLET_JSON = "{\n" +
            "\t\"type\": \"google_pay\",\n" +
            "\t\"dynamic_last4\": \"1234\",\n" +
            "\t\"google_pay\": {}\n" +
            "}";

    private static final String SAMSUNG_PAY_WALLET_JSON = "{\n" +
            "\t\"type\": \"samsung_pay\",\n" +
            "\t\"dynamic_last4\": \"1234\",\n" +
            "\t\"samsung_pay\": {}\n" +
            "}";

    @Test
    public void testCreateVisaCheckoutWallet() throws JSONException {
        final JSONObject walletJson = new JSONObject(VISA_WALLET_JSON);
        final Wallet wallet = new WalletFactory().create(walletJson);
        assertTrue(wallet instanceof VisaCheckoutWallet);
    }

    @Test
    public void testCreateMasterpassWallet() throws JSONException {
        final JSONObject walletJson = new JSONObject(MASTERPASS_WALLET_JSON);
        final Wallet wallet = new WalletFactory().create(walletJson);
        assertTrue(wallet instanceof MasterpassWallet);
    }

    @Test
    public void testCreateAmexExpressCheckoutWallet() throws JSONException {
        final JSONObject walletJson = new JSONObject(AMEX_EXPRESS_CHECKOUT_WALLET_JSON);
        final Wallet wallet = new WalletFactory().create(walletJson);
        assertTrue(wallet instanceof AmexExpressCheckoutWallet);
    }

    @Test
    public void testCreateApplePayWallet() throws JSONException {
        final JSONObject walletJson = new JSONObject(APPLE_PAY_WALLET_JSON);
        final Wallet wallet = new WalletFactory().create(walletJson);
        assertTrue(wallet instanceof ApplePayWallet);
    }

    @Test
    public void testCreateGooglePayWallet() throws JSONException {
        final JSONObject walletJson = new JSONObject(GOOGLE_PAY_WALLET_JSON);
        final Wallet wallet = new WalletFactory().create(walletJson);
        assertTrue(wallet instanceof GooglePayWallet);
    }

    @Test
    public void testCreateSamsungPayWallet() throws JSONException {
        final JSONObject walletJson = new JSONObject(SAMSUNG_PAY_WALLET_JSON);
        final Wallet wallet = new WalletFactory().create(walletJson);
        assertTrue(wallet instanceof SamsungPayWallet);
    }

    @Test
    public void testParcelable_shouldBeEqualAfterParcel() throws JSONException {
        final WalletFactory walletFactory = new WalletFactory();

        final SamsungPayWallet samsungPayWallet =
                (SamsungPayWallet) walletFactory.create(new JSONObject(SAMSUNG_PAY_WALLET_JSON));
        assertNotNull(samsungPayWallet);
        final Parcel samsungWalletParcel = Parcel.obtain();
        samsungPayWallet.writeToParcel(samsungWalletParcel, samsungPayWallet.describeContents());
        samsungWalletParcel.setDataPosition(0);
        final SamsungPayWallet parcelSamsungWallet =
                SamsungPayWallet.CREATOR.createFromParcel(samsungWalletParcel);
        assertEquals(samsungPayWallet, parcelSamsungWallet);

        final VisaCheckoutWallet visaWallet =
                (VisaCheckoutWallet) walletFactory.create(new JSONObject(VISA_WALLET_JSON));
        assertNotNull(visaWallet);
        final Parcel visaParcel = Parcel.obtain();
        visaWallet.writeToParcel(visaParcel, visaWallet.describeContents());
        visaParcel.setDataPosition(0);
        final VisaCheckoutWallet parcelVisaWallet =
                VisaCheckoutWallet.CREATOR.createFromParcel(visaParcel);
        assertEquals(visaWallet, parcelVisaWallet);
    }
}

File: stripe/src/test/java/com/stripe/android/model/wallets/WalletFactoryTest.kt
package com.stripe.android.model.wallets

import com.stripe.android.utils.ParcelUtils
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import org.json.JSONObject
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class WalletFactoryTest {

    @Test
    fun testCreateVisaCheckoutWallet() {
        val wallet = WalletFactory().create(VISA_WALLET_JSON)
        assertTrue(wallet is VisaCheckoutWallet)
    }

    @Test
    fun testCreateMasterpassWallet() {
        val wallet = WalletFactory().create(MASTERPASS_WALLET_JSON)
        assertTrue(wallet is MasterpassWallet)
    }

    @Test
    fun testCreateAmexExpressCheckoutWallet() {
        val wallet = WalletFactory().create(AMEX_EXPRESS_CHECKOUT_WALLET_JSON)
        assertTrue(wallet is AmexExpressCheckoutWallet)
    }

    @Test
    fun testCreateApplePayWallet() {
        val wallet = WalletFactory().create(APPLE_PAY_WALLET_JSON)
        assertTrue(wallet is ApplePayWallet)
    }

    @Test
    fun testCreateGooglePayWallet() {
        val wallet = WalletFactory().create(GOOGLE_PAY_WALLET_JSON)
        assertTrue(wallet is GooglePayWallet)
    }

    @Test
    fun testCreateSamsungPayWallet() {
        val wallet = WalletFactory().create(SAMSUNG_PAY_WALLET_JSON)
        assertTrue(wallet is SamsungPayWallet)
    }

    @Test
    fun testParcelable_shouldBeEqualAfterParcel() {
        val walletFactory = WalletFactory()

        val samsungPayWallet =
            walletFactory.create(SAMSUNG_PAY_WALLET_JSON) as SamsungPayWallet
        assertEquals(
            samsungPayWallet,
            ParcelUtils.create(samsungPayWallet, SamsungPayWallet.CREATOR)
        )

        val visaWallet =
            walletFactory.create(VISA_WALLET_JSON) as VisaCheckoutWallet
        assertEquals(
            visaWallet,
            ParcelUtils.create(visaWallet, VisaCheckoutWallet.CREATOR)
        )
    }

    companion object {
        private val VISA_WALLET_JSON = JSONObject(
            """
            {
                "type": "visa_checkout",
                "dynamic_last4": "1234",
                "visa_checkout": {
                    "billing_address": {
                        "city": "San Francisco",
                        "country": "US",
                        "line1": "510 Townsend St",
                        "postal_code": "94103",
                        "state": "CA"
                    },
                    "email": "me@example.com",
                    "name": "John Doe",
                    "shipping_address": {
                        "city": "San Francisco",
                        "country": "US",
                        "line1": "1355 Market St",
                        "postal_code": "94103",
                        "state": "CA"
                    }
                }
            }
            """.trimIndent()
        )

        private val MASTERPASS_WALLET_JSON = JSONObject(
            """
            {
                "type": "master_pass",
                "dynamic_last4": "1234",
                "master_pass": {
                    "billing_address": {
                        "city": "San Francisco",
                        "country": "US",
                        "line1": "510 Townsend St",
                        "postal_code": "94103",
                        "state": "CA"
                    },
                    "email": "me@example.com",
                    "name": "John Doe",
                    "shipping_address": {
                        "city": "San Francisco",
                        "country": "US",
                        "line1": "1355 Market St",
                        "postal_code": "94103",
                        "state": "CA"
                    }
                }
            }
            """.trimIndent()
        )

        private val AMEX_EXPRESS_CHECKOUT_WALLET_JSON = JSONObject(
            """
            {
                "type": "amex_express_checkout",
                "dynamic_last4": "1234",
                "amex_express_checkout": {}
            }
            """.trimIndent()
        )

        private val APPLE_PAY_WALLET_JSON = JSONObject(
            """
            {
                "type": "apple_pay",
                "dynamic_last4": "1234",
                "apple_pay": {}
            }
            """.trimIndent()
        )

        private val GOOGLE_PAY_WALLET_JSON = JSONObject(
            """
            {
                "type": "google_pay",
                "dynamic_last4": "1234",
                "google_pay": {}
            }
            """.trimIndent()
        )

        private val SAMSUNG_PAY_WALLET_JSON = JSONObject(
            """
            {
                "type": "samsung_pay",
                "dynamic_last4": "1234",
                "samsung_pay": {}
            }
            """.trimIndent()
        )
    }
}
